doc/design/migration_to_python.md

```diff
@@ -0,0 +1,479 @@
+
+The section [The Proposed Architecture](#the-proposed-architecture) illustrates an outline design about how the migration will be done.
+
+The last two sections explain how the proposed architecture solves the problem of the current architecture, as well as the preparations to be noted.
+
+## Flaws In The Current Architecture
+
+For SQLFlow, migrating to python might mean rewriting lots of existing golang code in python. Why should we migrate the golang *ML code* to python? Is is necessary? Is it the right time?
+
+After demonstrating the flaws of the current architecture in the following scenarios, the answer will be obvious.
+
+### Scenario 1: An ML Expert Wants A New Column Type
+
+Feature engineering is a very important part of machine learning. SQLFlow supports feature engineering via the `COLUMN` clause. The current `COLUMN` clause of SQLFlow is based on `tf.feature_column` and works only for TensorFlow. In the next version, the SQLFlow `COLUMN` clause will support more comprehensive feature columns for TensorFlow, PyTorch and XGBoost. 
+
+Suppose the user wants a `image_embedding_column ` that converts a input image into embedding for a recommender system. Adding such a new column in the current architecture requires the following main steps:
+
+1. The python part:
+   1. Implement the column in a python file. 
+2. The golang part:
+   1. Add a `ImageEmbeddingColumn` struct in [pkg/ir/feature_column.go](https://github.com/sql-machine-learning/sqlflow/tree/400c691470c6503393453d47856913df3365503e/pkg/ir/feature_column.go).
+   2. Add a `parseImageEmbeddingColumn` function in [pkg/sql/ir_generator.go](https://github.com/sql-machine-learning/sqlflow/tree/400c691470c6503393453d47856913df3365503e/pkg/sql/ir_generator.go).
+   3. Add a `switch` statement in the function `initColumnMap` in [pkg/step/feature/derivation.go](https://github.com/sql-machine-learning/sqlflow/tree/400c691470c6503393453d47856913df3365503e/pkg/step/feature/derivation.go).
+   4. Add a `case` in the large `switch` statement in function `generateFeatureColumnCode` in [pkg/sql/codegen/tensorflow/codegen.go](https://github.com/sql-machine-learning/sqlflow/tree/400c691470c6503393453d47856913df3365503e/pkg/sql/codegen/tensorflow/codegen.go). This step generates a snippet of python code with simple text substitution.
+   5. Check inputs for `image_embedding_column ` to avoid possible run-time exception from the python part to ensure user experience.
+3. Test and debug: a bug can be in the python file, in `pkg/ir/feature_column.go`, in `pkg/sql/ir_generator.go`, in `pkg/step/feature/derivation.go`, in `pkg/sql/codegen/tensorflow/codegen.go`, or in the python code text substitution template.
+
+See https://github.com/sql-machine-learning/sqlflow/pull/1897 and https://github.com/sql-machine-learning/sqlflow/pull/1901 for an real-world example.
+
+### Scenario 2: An ML Expert Wants To Contribute A New Type Of Model
```

Comment:
Same as above. Yet, after this refactor, we can make SQLFlow easier to do continuously evolving.